File: app/src/main/java/com/example/music/hi/tab/bottom/HiTabBottomLayout.java
package com.example.music.hi.tab.bottom;

import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.lib_hi.util.HiDisplayUtil;
import com.example.music.R;
import com.example.music.hi.tab.common.IHiTabLayout;


import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class HiTabBottomLayout extends FrameLayout implements IHiTabLayout<HiTabItemBottom, HiTabBottomInfo<?>> {
    private List<OnTabSelectedListener<HiTabBottomInfo<?>>> tabSelectedChangeListeners = new ArrayList<>();
    private HiTabBottomInfo<?> selectedInfo;
    private float bottomAlpha = 1f;
    //TabBottom高度
    private float tabBottomHeight = 50;
    //TabBottom的头部线条高度
    private float bottomLineHeight = 0.5f;
    //TabBottom的头部线条颜色
    private String bottomLineColor = "#dfe0e1";
    private List<HiTabBottomInfo<?>> infoList;

    public HiTabBottomLayout(@NonNull Context context) {
        this(context, null);
    }

    public HiTabBottomLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public HiTabBottomLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public HiTabItemBottom findTab(@NonNull HiTabBottomInfo<?> info) {
        ViewGroup ll = findViewWithTag(TAG_TAB_BOTTOM);
        for (int i = 0; i < ll.getChildCount(); i++) {
            View child = ll.getChildAt(i);
            if (child instanceof HiTabItemBottom) {
                HiTabItemBottom tab = (HiTabItemBottom) child;
                if (tab.getTabInfo() == info) {
                    return tab;
                }
            }
        }
        return null;
    }

    @Override
    public void addTabSelectedChangeListener(OnTabSelectedListener<HiTabBottomInfo<?>> listener) {
        tabSelectedChangeListeners.add(listener);
    }

    @Override
    public void defaultSelected(@NonNull HiTabBottomInfo<?> defaultInfo) {
        onSelected(defaultInfo);
    }

    private static final String TAG_TAB_BOTTOM = "TAG_TAB_BOTTOM";

    @Override
    public void inflateInfo(@NonNull List<HiTabBottomInfo<?>> infoList) {
        if (infoList.isEmpty()) return;
        this.infoList = infoList;
        // 移除之前界面上的内容，第0个不移除，表示上方内容容器
        for (int i = getChildCount() - 1; i > 0; i--) {
            removeViewAt(i);
        }
        //清除之前添加的HiTabBottom listener，Tips：Java foreach 不嫩在遍历过程中对集合增删改
        Iterator<OnTabSelectedListener<HiTabBottomInfo<?>>> iterator = tabSelectedChangeListeners.iterator();
        while (iterator.hasNext()) {
            if (iterator.next() instanceof HiTabItemBottom) {
                iterator.remove();
            }
        }
        this.selectedInfo = new HiTabBottomInfo<String>();

        // 添加背景
        addBackground();

        // 创建承载Item的FrameLayout，并添加Item
        int height = HiDisplayUtil.dp2px(tabBottomHeight, getResources());
        FrameLayout ll = new FrameLayout(getContext());
        ll.setTag(TAG_TAB_BOTTOM);
        int width = HiDisplayUtil.getDisplayWidthInPx(getContext()) / infoList.size();
        for (int i = 0; i < infoList.size(); i++) {
            HiTabBottomInfo<?> info = infoList.get(i);
            LayoutParams params = new LayoutParams(width, height);
            params.gravity = Gravity.BOTTOM;
            params.leftMargin = i * width;

            HiTabItemBottom tabBottom = new HiTabItemBottom(getContext());
            tabSelectedChangeListeners.add(tabBottom);
            tabBottom.setHiTabInfo(info);
            ll.addView(tabBottom, params);
            tabBottom.setOnClickListener(v -> onSelected(info));
        }
        LayoutParams llParams = new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
        llParams.gravity = Gravity.BOTTOM;

        // 加条线
        addBottomLine();

        // 添加承载Item的FrameLayout
        addView(ll, llParams);
    }

    /**
     * 当选中的bottomItem更改时，遍历所有监听方法，对被选中的Item进行回调
     */
    private void onSelected(@NonNull HiTabBottomInfo<?> nextInfo) {
        for (OnTabSelectedListener<HiTabBottomInfo<?>> listener : tabSelectedChangeListeners) {
            listener.onTabSelectedChange(infoList.indexOf(nextInfo), selectedInfo, nextInfo);
        }
        this.selectedInfo = nextInfo;
    }

    private void addBottomLine() {
        View bottomLine = new View(getContext());
        bottomLine.setBackgroundColor(Color.parseColor(bottomLineColor));

        LayoutParams bottomLineParams =
                new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, HiDisplayUtil.dp2px(bottomLineHeight, getResources()));
        bottomLineParams.gravity = Gravity.BOTTOM;
        bottomLineParams.bottomMargin = HiDisplayUtil.dp2px(tabBottomHeight - bottomLineHeight, getResources());
        addView(bottomLine, bottomLineParams);
        bottomLine.setAlpha(bottomAlpha);
    }

    private void addBackground() {
        View view = LayoutInflater.from(getContext()).inflate(R.layout.hi_bottom_layout_bg, null);
        LayoutParams params =
                new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, HiDisplayUtil.dp2px(tabBottomHeight, getResources()));
        params.gravity = Gravity.BOTTOM;
        addView(view, params);
        view.setAlpha(bottomAlpha);
    }

    public void setBottomAlpha(float bottomAlpha) {
        this.bottomAlpha = bottomAlpha;
    }

    public void setTabBottomHeight(float tabBottomHeight) {
        this.tabBottomHeight = tabBottomHeight;
    }

    public void setBottomLineHeight(float bottomLineHeight) {
        this.bottomLineHeight = bottomLineHeight;
    }

    public void setBottomLineColor(String bottomLineColor) {
        this.bottomLineColor = bottomLineColor;
    }
}

File: app/src/main/java/com/example/music/hi/tab/bottom/HiTabBottomLayout.kt
package com.example.music.hi.tab.bottom

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AbsListView
import android.widget.FrameLayout
import android.widget.ScrollView
import androidx.annotation.NonNull
import androidx.core.view.forEach
import androidx.recyclerview.widget.RecyclerView
import com.example.lib_hi.util.HiDisplayUtil
import com.example.lib_hi.util.HiViewUtil
import com.example.music.R
import com.example.music.hi.tab.common.IHiTabLayout

class HiTabBottomLayout @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0)
    : FrameLayout(context, attrs, defStyleAttr), IHiTabLayout<HiTabItemBottom, HiTabBottomInfo<*>> {

    var selectInfo: HiTabBottomInfo<*> = HiTabBottomInfo<String>()
    var bottomAlpha = 1f
    var tabBottomHeight = 50f
    var bottomLineHeight = 0.5f
    var bottomLineColor = "#dfe0e1"
    var infoList: List<HiTabBottomInfo<*>>? = null
    private val tabSelectedChangeListeners: MutableList<IHiTabLayout.OnTabSelectedListener<HiTabBottomInfo<*>>> = ArrayList()

    companion object {
        const val TAG_TAB_BOTTOM = "TAG_TAB_BOTTOM"
    }

    override fun findTab(data: HiTabBottomInfo<*>): HiTabItemBottom? {
        val ll = findViewWithTag<ViewGroup>(TAG_TAB_BOTTOM)
        ll.forEach {
            if (it is HiTabItemBottom && it.tabInfo == data) {
                return it
            }
        }
        return null
    }

    override fun addTabSelectedChangeListener(listener: IHiTabLayout.OnTabSelectedListener<HiTabBottomInfo<*>>) {
        this.tabSelectedChangeListeners.add(listener)
    }

    override fun defaultSelected(defaultInfo: HiTabBottomInfo<*>) {
        onSelected(defaultInfo)
        this.selectInfo = defaultInfo
    }

    override fun inflateInfo(infoList: List<HiTabBottomInfo<*>>) {
        this.infoList = infoList
        // 移除之前界面上的内容，第0个不移除，表示上方内容容器
        for (i in childCount - 1 downTo 1) {
            removeViewAt(i)
        }
        this.tabSelectedChangeListeners.forEach {
            if (it is HiTabItemBottom) {
                tabSelectedChangeListeners.remove(it)
            }
        }

        // 添加背景
        addBackground()

        // 创建承载Item的FrameLayout ll，并逐个添加tabItemBottom
        val height = HiDisplayUtil.dp2px(tabBottomHeight, resources)
        val ll = FrameLayout(context)
        ll.tag = TAG_TAB_BOTTOM
        val width = HiDisplayUtil.getDisplayWidthInPx(context) / infoList.size
        this.infoList?.forEachIndexed { i, info ->
            val params = LayoutParams(width, height, Gravity.BOTTOM)
            params.leftMargin = i * width
            val tabItemBottom = HiTabItemBottom(context)
            tabItemBottom.setHiTabInfo(info)
            tabItemBottom.setOnClickListener { onSelected(info) }
            tabSelectedChangeListeners.add(tabItemBottom)
            ll.addView(tabItemBottom, params)
        }
        val llParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT, Gravity.BOTTOM)
        addView(ll, llParams)

        // 加条线
        addBottomLine()

        // 修复最后一个view的展示
        fixContentView()
    }

    private fun addBackground() {
        val view = LayoutInflater.from(context).inflate(R.layout.hi_bottom_layout_bg, null)
        val params = LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            HiDisplayUtil.dp2px(this.tabBottomHeight, resources),
            Gravity.BOTTOM)
        addView(view, params)
        view.alpha = this.bottomAlpha
    }

    private fun addBottomLine() {
        val bottomLine = View(context)
        bottomLine.setBackgroundColor(Color.parseColor(this.bottomLineColor))
        val bottomLineParams = LayoutParams(
            LayoutParams.MATCH_PARENT,
            HiDisplayUtil.dp2px(this.bottomLineHeight, resources),
            Gravity.BOTTOM)
        addView(bottomLine, bottomLineParams)
        bottomLine.alpha = this.bottomAlpha
    }

    private fun onSelected(@NonNull nextInfo: HiTabBottomInfo<*>) {
        tabSelectedChangeListeners.forEach {
            infoList?.indexOf(nextInfo)?.let { it1 -> it.onTabSelectedChange(it1, selectInfo, nextInfo) }
        }
        this.selectInfo = nextInfo
    }

    private fun fixContentView() {
        if (getChildAt(0) !is ViewGroup) return
        val rootView : ViewGroup = getChildAt(0) as ViewGroup
        val targetView = HiViewUtil.findTypeView(rootView, RecyclerView::class.java)
                ?: HiViewUtil.findTypeView(rootView, ScrollView::class.java)
                ?: HiViewUtil.findTypeView(rootView, AbsListView::class.java)
        targetView?.setPadding(0, 0, 0, HiDisplayUtil.dp2px(this.tabBottomHeight, resources))
        targetView?.clipToPadding = false
    }
}